cpp/DEVELOPER_GUIDE.md

```diff
@@ -9,6 +9,85 @@ cuML is thread safe so its functions can be called from multiple host threads if
 
 The implementation of cuML is single threaded.
 
+## Public cuML interface
+### Motivation
+The public cuML interface is stateless for two main reasons:
```

Comment:
We should distinguish between the stateless and stateful convenience APIs here, which are also public. 

`base public API` or `core public API` seem like good choices here.

Reply:
Done. Also added a small terminology section to be more explicit about the terms used.